### PR TITLE
cpu: aarch64: reorder: support jit-ed blk_reorder

### PR DESCRIPTION
**Pull request by kawakami-k**
# Description

This PR adds jit-ed blk_reorder for AArch64.

# Checklist

## General

- [x] Do all unit and benchdnn tests (`make test` and `make test_benchdnn_*`) pass locally for each commit?
- [x] Have you formatted the code using clang-format?

### CODE REVIEW

File: src/cpu/aarch64/jit_uni_reorder.hpp [hunk 1]
```diff
@@ -1,6 +1,7 @@
 /*******************************************************************************
-* Copyright 2018-2020 Intel Corporation
-* Copyright 2020 FUJITSU LIMITED
+* Copyright 2018-2022 Intel Corporation
+* Copyright 2020-2022 FUJITSU LIMITED
+* Copyright 2022 Arm Ltd. and affiliates
```
Comment (cfRod) [inline src/cpu/aarch64/jit_uni_reorder.hpp:4]: Hi Kawakami,

Thanks for your PR!

I see that you have included the Arm Licence header here and it jit_reorder.cpp.  I wanted to check if it is because my work towards the reorder implementation in this PR: https://github.com/oneapi-src/oneDNN/pull/1475 and if yes, I think it should be included as a separate commit and also list me as co-author?

Comment (kawakami-k) [reply src/cpu/aarch64/jit_uni_reorder.hpp:4]: @cfRod 
Thank you for the comment and sorry for dropping out your commit log! 
I rebuilt the commit log of this PR. 

image

**Comment by dzarukin** [comment]
Hi @kawakami-k, could you, please, update this PR on top of latest master? Thank you!

**Comment by kawakami-k** [comment]
> Hi @kawakami-k, could you, please, update this PR on top of latest master? Thank you!

@dzarukin 
Thank you for the comment. I updated.

File: src/cpu/aarch64/jit_uni_reorder.hpp (hunk 1 above)
Comment (cfRod) [reply src/cpu/aarch64/jit_uni_reorder.hpp:4]: Thanks!

**Comment by cfRod** [comment]
I have done some testing of your PRs #1483 and #1485 applied on HEAD [8376239](https://github.com/oneapi-src/oneDNN/commit/837623961e655753c96fa79ae5d6887bf4912da0) on SVE -256 and SVE-512
On sve-256, everything looks fine. All ctests passing,
```
     Start 96: test_regression
96/97 Test #96: test_regression ..................................   Passed    0.00 sec
      Start 97: noexcept-cpp
97/97 Test #97: noexcept-cpp .....................................   Passed    0.00 sec

100% tests passed, 0 tests failed out of 97

Total Test time (real) =  90.78 sec
```
However there is one ctest case failing on SVE-512:  Convolution_SimpleSmall_Blocked16_Tail_eltwise_CPU
[](url)
[ctest-sve512.log](https://github.com/oneapi-src/oneDNN/files/9910688/ctest-sve512.log)

**Comment by cfRod** [comment]
Further update:

I looked at the failing ctest case and I have a benchdnn command based on the problem description from oneDNN verbose..
I remove the eltwise-post ops and there is an error on SVE-512 for check_zero_padding failed.

on SVE-256
```
DNNL_VERBOSE=1 OMP_NUM_THREADS=1 ./tests/benchdnn/benchdnn --conv --stag=aBcd16b --wtag=Abcde16a --dtag=aBcd16b g47mb1_ic47oc47_ih20oh20kh3sh1dh0ph1_iw20ow20kw3sw1dw0pw1
onednn_verbose,info,oneDNN v2.7.0 (commit 837623961e655753c96fa79ae5d6887bf4912da0)
onednn_verbose,info,cpu,runtime:OpenMP,nthr:1
onednn_verbose,info,cpu,isa:AArch64 SVE (256 bits)
onednn_verbose,info,gpu,runtime:none
onednn_verbose,info,prim_template:operation,engine,primitive,implementation,prop_kind,memory_descriptors,attributes,auxiliary,problem_desc,exec_time
onednn_verbose,exec,cpu,reorder,jit:uni,undef,src_f32::blocked:abcd:f0 dst_f32:p:blocked:aBcd16b:f0,,,1x47x20x20,0.0109863
onednn_verbose,exec,cpu,reorder,jit:uni,undef,src_f32::blocked:abcde:f0 dst_f32:p:blocked:Abcde16a:f0,,,47x1x1x3x3,0.000976562
onednn_verbose,exec,cpu,reorder,simple:any,undef,src_f32::blocked:a:f0 dst_f32::blocked:a:f0,,,47,0.032959
onednn_verbose,exec,cpu,convolution,ref:any,forward_training,src_f32:p:blocked:aBcd16b:f0 wei_f32:p:blocked:Abcde16a:f0 bia_f32::blocked:a:f0 dst_f32:p:blocked:aBcd16b:f0,,alg:convolution_direct,g47mb1_ic47oc47_ih20oh20kh3sh1dh0ph1_iw20ow20kw3sw1dw0pw1,3.29102
onednn_verbose,exec,cpu,reorder,jit:uni,undef,src_f32:p:blocked:aBcd16b:f0 dst_f32::blocked:abcd:f0,,,1x47x20x20,0.0100098
0:PASSED __REPRO: --conv --stag=aBcd16b --wtag=Abcde16a --dtag=aBcd16b g47mb1ic47ih20oc47oh20kh3ph1
tests:1 passed:1 skipped:0 mistrusted:0 unimplemented:0 invalid_arguments:0 failed:0 listed:0
total compute_ref: sum(s):0.00 
```
on SVE-512
```
DNNL_VERBOSE=1 OMP_NUM_THREADS=1 ./tests/benchdnn/benchdnn --conv --stag=aBcd16b --wtag=Abcde16a --dtag=aBcd16b g47mb1_ic47oc47_ih20oh20kh3sh1dh0ph1_iw20ow20kw3sw1dw0pw1
onednn_verbose,info,oneDNN v2.7.0 (commit 837623961e655753c96fa79ae5d6887bf4912da0)
onednn_verbose,info,cpu,runtime:OpenMP,nthr:1
onednn_verbose,info,cpu,isa:AArch64 SVE (512 bits)
onednn_verbose,info,gpu,runtime:none
onednn_verbose,info,prim_template:operation,engine,primitive,implementation,prop_kind,memory_descriptors,attributes,auxiliary,problem_desc,exec_time
 jit_uni_reorder_t::execute
onednn_verbose,exec,cpu,reorder,jit:uni,undef,src_f32::blocked:abcd:f0 dst_f32:p:blocked:aBcd16b:f0,,,1x47x20x20,0.0620117
 jit_uni_reorder_t::execute
onednn_verbose,exec,cpu,reorder,jit:uni,undef,src_f32::blocked:abcde:f0 dst_f32:p:blocked:Abcde16a:f0,,,47x1x1x3x3,0.019043
onednn_verbose,exec,cpu,reorder,simple:any,undef,src_f32::blocked:a:f0 dst_f32::blocked:a:f0,,,47,0.00585938
onednn_verbose,exec,cpu,convolution,jit_dw:sve_512,forward_training,src_f32:p:blocked:aBcd16b:f0 wei_f32:p:blocked:Abcde16a:f0 bia_f32::blocked:a:f0 dst_f32:p:blocked:aBcd16b:f0,,alg:convolution_direct,g47mb1_ic47oc47_ih20oh20kh3sh1dh0ph1_iw20ow20kw3sw1dw0pw1,0.0219727
[12815][arg:17][0,47,0,0,0,0] fp:  0.f dt:     -nan 
[12831][arg:17][0,47,0,1,0,0] fp:  0.f dt:     -nan 
[12847][arg:17][0,47,0,2,0,0] fp:  0.f dt:     -nan 
[12863][arg:17][0,47,0,3,0,0] fp:  0.f dt:     -nan 
[12879][arg:17][0,47,0,4,0,0] fp:  0.f dt:     -nan 
[12895][arg:17][0,47,0,5,0,0] fp:  0.f dt:     -nan 
[12911][arg:17][0,47,0,6,0,0] fp:  0.f dt:     -nan 
[12927][arg:17][0,47,0,7,0,0] fp:  0.f dt:     -nan 
[12943][arg:17][0,47,0,8,0,0] fp:  0.f dt:     -nan 
@@@ [arg:17] check_zero_padding failed
 jit_uni_reorder_t::execute
onednn_verbose,exec,cpu,reorder,jit:uni,undef,src_f32:p:blocked:aBcd16b:f0 dst_f32::blocked:abcd:f0,,,1x47x20x20,0.0490723
[12800][DST][0:32:0:0] exp_f32:     968 exp:     968 got:    -nan diff:     nan rdiff:     nan
[12801][DST][0:32:0:1] exp_f32:      22 exp:      22 got:    -nan diff:     nan rdiff:     nan
[12802][DST][0:32:0:2] exp_f32:     204 exp:     204 got:    -nan diff:     nan rdiff:     nan
[12803][DST][0:32:0:3] exp_f32:    1152 exp:    1152 got:    -nan diff:     nan rdiff:     nan
[12804][DST][0:32:0:4] exp_f32:       2 exp:       2 got:    -nan diff:     nan rdiff:     nan
[12805][DST][0:32:0:5] exp_f32:     156 exp:     156 got:    -nan diff:     nan rdiff:     nan
[12806][DST][0:32:0:6] exp_f32:    1336 exp:    1336 got:    -nan diff:     nan rdiff:     nan
[12807][DST][0:32:0:7] exp_f32:     -18 exp:     -18 got:    -nan diff:     nan rdiff:     nan
[12808][DST][0:32:0:8] exp_f32:    -462 exp:    -462 got:    -nan diff:     nan rdiff:     nan
0:FAILED (errors:6000 total:18800) __REPRO: --conv --stag=aBcd16b --wtag=Abcde16a --dtag=aBcd16b g47mb1ic47ih20oc47oh20kh3ph1
tests:1 passed:0 skipped:0 mistrusted:0 unimplemented:0 invalid_arguments:0 failed:1 listed:0
total compute_ref: sum(s):0.01 
```

**Comment by kawakami-k** [comment]
> @kawakami-k, I have done some testing of your PRs #1483 and #1485 applied on HEAD [8376239](https://github.com/oneapi-src/oneDNN/commit/837623961e655753c96fa79ae5d6887bf4912da0) on SVE -256 and SVE-512 On sve-256, everything looks fine. All ctests passing,
> 
> ```
>      Start 96: test_regression
> 96/97 Test #96: test_regression ..................................   Passed    0.00 sec
>       Start 97: noexcept-cpp
> 97/97 Test #97: noexcept-cpp .....................................   Passed    0.00 sec
> 
> 100% tests passed, 0 tests failed out of 97
> 
> Total Test time (real) =  90.78 sec
> ```
> 
> However there is one ctest case failing on SVE-512: Convolution_SimpleSmall_Blocked16_Tail_eltwise_CPU. See attached log: [ctest-sve512.log](https://github.com/oneapi-src/oneDNN/files/9910722/ctest-sve512.log)

@cfRod 
Thank you! `jit_uni_dw_conv_kernel_utils.hpp` may be degraded. I'm checking it.

**Review by dzarukin — APPROVED** [review]

**Comment by dzarukin** [comment]
Thank you for the contribution!

**Comment by kawakami-k** [comment]
@cfRod 

`./tests/benchdnn/benchdnn --conv --stag=aBcd16b --wtag=Abcde16a --dtag=aBcd16b g47mb1_ic47oc47_ih20oh20kh3sh1dh0ph1_iw20ow20kw3sw1dw0pw1` will be fixed by 
https://github.com/oneapi-src/oneDNN/pull/1489

Thank you!

**Comment by cfRod** [comment]
Thanks @kawakami-k !
